File: Introibo/Data/MassProper.swift
import Foundation

struct MassProper: Identifiable, Decodable, Hashable {
    let slug: String
    let title: String
    let english: String
    let rank: Int
    let color: String
    let season: String?
    let introit: ProperText
    let collect: ProperText
    let epistle: ProperReading
    let gradual: ProperText?
    let alleluia: ProperText?
    let tract: ProperText?
    let sequence: ProperText?
    let gospel: ProperReading
    let offertory: ProperText
    let secret: ProperText
    let communion: ProperText
    let postcommunion: ProperText
    let preface: String?
    // Optional rubric hints from DO data (rule.gloria, rule.credo).
    // When present, override the rank-based heuristic.
    var glorOverride: Bool? = nil
    var credoOverride: Bool? = nil
    var related: [RelatedLink]? = nil

    var id: String { slug }

    /// English feast/feria name, resolved lazily from the translation map.
    /// Falls back to the stored `english` field (which is the Latin title)
    /// if no translation exists.
    var englishTitle: String {
        ContentStore.shared.ordoNameEnglish(title) ?? english
    }
}

struct ProperText: Decodable, Hashable {
    let lat: String
    let eng: String
    var ref: String? = nil
}

struct ProperReading: Decodable, Hashable {
    let ref: String
    let lat: String
    let eng: String
}

File: android/app/src/main/java/com/lampstandhq/introibo/data/model/MassProper.kt
package com.lampstandhq.introibo.data.model

import com.lampstandhq.introibo.data.content.ContentStore
import com.lampstandhq.introibo.data.links.RelatedLink
import kotlinx.serialization.Serializable

/**
 * Matches propers.json — Mass propers for a given day/feast.
 */
@Serializable
data class MassProper(
    val slug: String,
    val title: String,
    val english: String,
    val rank: Int,
    val color: String,
    val season: String? = null,
    val introit: ProperText,
    val collect: ProperText,
    val epistle: ProperReading,
    val gradual: ProperText? = null,
    val alleluia: ProperText? = null,
    val tract: ProperText? = null,
    val sequence: ProperText? = null,
    val gospel: ProperReading,
    val offertory: ProperText,
    val secret: ProperText,
    val communion: ProperText,
    val postcommunion: ProperText,
    val preface: String? = null,
    // Optional rubric hints from DO data (rule.gloria, rule.credo)
    val glorOverride: Boolean? = null,
    val credoOverride: Boolean? = null,
    val related: List<RelatedLink>? = null,
) {
    /** English feast/feria name, resolved lazily from the translation map. */
    val englishTitle: String
        get() = ContentStore.ordoNameEnglish(title) ?: english
}

@Serializable
data class ProperText(
    val lat: String,
    val eng: String,
    val ref: String? = null,
)

@Serializable
data class ProperReading(
    val ref: String,
    val lat: String,
    val eng: String,
)
